src/cascadia/WindowsTerminal/WindowThread.cpp

```diff
@@ -0,0 +1,143 @@
+
+WindowThread::WindowThread(const winrt::TerminalApp::AppLogic& logic,
+                           winrt::Microsoft::Terminal::Remoting::WindowRequestedArgs args,
+                           winrt::Microsoft::Terminal::Remoting::WindowManager manager,
+                           winrt::Microsoft::Terminal::Remoting::Peasant peasant) :
+    _peasant{ peasant },
+    _appLogic{ logic },
+    _args{ args },
+    _manager{ manager }
+{
+    // DO NOT start the AppHost here in the ctor, as that will start XAML on the wrong thread!
+}
+
+void WindowThread::Start()
+{
+    _thread = std::thread([this]() {
+        // Start the AppHost HERE, on the actual thread we want XAML to run on
+        _host = std::make_unique<::AppHost>(_appLogic,
+                                            _args,
+                                            _manager,
+                                            _peasant);
+        _host->UpdateSettingsRequested([this]() { _UpdateSettingsRequestedHandlers(); });
+        // Enter the main window loop.
+        const auto exitCode = WindowProc();
+        _host = nullptr;
+
+        // !! LOAD BEARING !!
+        //
+        // Make sure to finish pumping all the messages for our thread here. We
+        // may think we're all done, but we're not quite. XAML needs more time
+        // to pump the remaining events through, even at the point we're
+        // exiting. So do that now. If you don't, then the last tab to close
+        // will never actually destruct the last tab / TermControl / ControlCore
+        // / renderer.
+        {
+            MSG msg = {};
+            while (PeekMessageW(&msg, nullptr, 0, 0, PM_REMOVE))
+            {
+                ::DispatchMessageW(&msg);
+            }
+        }
+
+        // Raising our Exited event might cause the app to teardown. In the
+        // highly unlikely case that the main thread got scheduled before the
+        // event handler returns, it might release the WindowThread while it's
+        // still holding an outstanding reference to its own thread. That would
+        // cause a std::terminate. Let's just avoid that by releasing our
+        // thread.
+        this->_thread.detach();
+        _ExitedHandlers(_peasant.GetID());
+
+        return exitCode;
+    });
+    LOG_IF_FAILED(SetThreadDescription(_thread.native_handle(), L"Window Thread"));
+}
+
+winrt::TerminalApp::TerminalWindow WindowThread::Logic()
+{
+    return _host->Logic();
+}
+
+static bool _messageIsF7Keypress(const MSG& message)
+{
+    return (message.message == WM_KEYDOWN || message.message == WM_SYSKEYDOWN) && message.wParam == VK_F7;
+}
+static bool _messageIsAltKeyup(const MSG& message)
+{
+    return (message.message == WM_KEYUP || message.message == WM_SYSKEYUP) && message.wParam == VK_MENU;
+}
+static bool _messageIsAltSpaceKeypress(const MSG& message)
+{
+    return message.message == WM_SYSKEYDOWN && message.wParam == VK_SPACE;
+}
+
+int WindowThread::WindowProc()

```

Comment:
nit: naming overload here could be confusing for future developers. This is not a _WndProc_ in the user32 sense; this is a _message pump_ or _message loop_